warehouse/oidc/services.py

```diff
@@ -0,0 +1,168 @@
+        timeout is in effect.
+
+        Returns the refreshed keyset, or the cached keyset if a timeout is
+        in effect.
+
+        Returns the cached keyset on any provider access or format errors.
+        """
+
+        # Fast path: we're in a cooldown from a previous refresh.
+        keys, timeout = self._get_keyset()
+        if timeout:
+            self.metrics.increment("warehouse.oidc.refresh_keyset.timeout")
+            return keys
+
+        oidc_url = f"{oidc.OIDC_PROVIDERS[self.provider]}/{oidc.WELL_KNOWN_OIDC_CONF}"
+
+        resp = requests.get(oidc_url)
+
+        # For whatever reason, an OIDC provider's configuration URL might be
+        # offline. We don't want to completely explode here, since other
+        # providers might still be online (and need updating), so we spit
+        # out an error and return None instead of raising.
+        if not resp.ok:
+            sentry_sdk.capture_message(
+                f"OIDC provider {self.provider} failed to return configuration: "
+                f"{oidc_url}"
+            )
+            return keys
+
+        oidc_conf = resp.json()
+        jwks_url = oidc_conf.get("jwks_uri")
+
+        # A valid OIDC configuration MUST have a `jwks_uri`, but we
+        # defend against its absence anyways.
+        if jwks_url is None:
+            sentry_sdk.capture_message(
+                f"OIDC provider {self.provider} is returning malformed "
+                "configuration (no jwks_uri)"
+            )
+            return keys
+
+        resp = requests.get(jwks_url)
+
+        # Same reasoning as above.
+        if not resp.ok:
+            sentry_sdk.capture_message(
+                f"OIDC provider {self.provider} failed to return JWKS JSON: "
+                f"{jwks_url}"
+            )
+            return keys
+
+        jwks_conf = resp.json()
+        keys = jwks_conf.get("keys")
+
+        # Another sanity test: an OIDC provider should never return an empty
+        # keyset, but there's nothing stopping them from doing so. We don't
+        # want to cache an empty keyset just in case it's a short-lived error,
+        # so we check here, error, and return the current cache instead.
+        if not keys:
+            sentry_sdk.capture_message(
+                f"OIDC provider {self.provider} returned JWKS but no keys"
+            )
+            return keys
+
+        keys = {key["kid"]: key for key in keys}
+        self._store_keyset(keys)
+
+        return keys
+
+    def get_key(self, key_id):
+        """
+        Return a JWK for the given key ID, or None if the key can't be found
+        in this provider's keyset.
+        """
+
+        keyset, _ = self._get_keyset()
+        if key_id not in keyset:
+            keyset = self._refresh_keyset()
+        if key_id not in keyset:
+            self.metrics.increment(
+                "warehouse.oidc.get_key.error", tags=[f"key_id:{key_id}"]
+            )
```

Comment:
cc @di Does this metric look reasonable to you? I used the same pattern as I've seen elsewhere in Warehouse (fixed metric ID, dynamic tag formatted as `key:value`).